### PR TITLE
WELD-2523 Allow to register additional bean defining annotations in W…

### PR DESCRIPTION
**Pull request by manovotn**
…eld SE.

### CODE REVIEW

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java [hunk 1]
```diff
@@ -923,6 +939,8 @@ protected Deployment createDeployment(ResourceLoader resourceLoader, CDI11Bootst
             .addAll(typeDiscoveryConfiguration.getKnownBeanDefiningAnnotations())
             // Add ThreadScoped manually as Weld SE doesn't support implicit bean archives without beans.xml
             .add(ThreadScoped.class)
+            // Add all custom BDA user registered via Weld.addBeanDefiningAnnotations()
```
[the remark this answers is not in the document]
Comment (manovotn) [reply environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java]: Hehe, acronym overload.
Will change that.

[the remark this answers is not in the document]
Comment (nziakova) [reply environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java]: "BDA" is still used in the test - maybe it should be updated too (at least in the comment where it does not hurt to have the full term)?

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java [hunk 2]
```diff
@@ -887,6 +890,22 @@ public Weld enableDevMode() {
...
+    /**
+     * Registers annotations which will be considered as bean defining annotations.
+     *
+     * NOTE - If used along with {@code <trim/>} bean archives and/or with Weld configuration key
+     * {@code org.jboss.weld.bootstrap.vetoTypesWithoutBeanDefiningAnnotation}, these annotations will be ignored.
+     *
+     * @param annotations annotations which will be considered as Bean Defining Annotations.
+     * @return self
+     */
+    public Weld addBeanDefiningAnnotations(Class<? extends Annotation>... annotations) {
```
Comment (nziakova) [inline environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java]: Does it make sense to document this, e.g. in 15.6.2. Implicit bean archive (there's a NOTE in that section listing what are bean defining annotations)?

Comment (manovotn) [reply environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java]: Chapter 15 is about EE while this feature only concerns SE.
But you are right that we might want to mention this, but maybe in `18.4.2. Bootstrapping CDI SE` instead?

**Comment by manovotn** [comment]
@nziakova @mkouba added docs as well, take a glance please

File: environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java (hunk 2 above)
Comment (nziakova) [reply environments/se/core/src/main/java/org/jboss/weld/environment/se/Weld.java]: Of course you are right, apologies for confusion.

File: docs/reference/src/main/asciidoc/environments.asciidoc [hunk 3]
```diff
@@ -723,6 +723,14 @@ Another approach is to use `Weld.property()` and `SeContainerInitializer.addProp
 
 NOTE: The bean discovery mode of `annotated` is supported from version 2.2.0.Final. Previous versions processed implicit bean archives in the same way as explicit bean archives.
 
+==== Extending Bean Defining Annotations
+
+If you are running with discovery mode `annotated`, then only classes with bean defining annotations will be picked up as beans.
+The set of these annotations is given by CDI but Weld SE allows you to expand it via `Weld.addBeanDefiningAnnotations(Class<? extends Annotation>... annotations)`.
+Any annotation added this via will be considered a bean defining annotation when performing discovery.
```
Comment (nziakova) [inline docs/reference/src/main/asciidoc/environments.asciidoc]: Look like a wrong word order: this via -> via this? But I don't like "via this" too much, maybe e.g. "this way" would be better?

Comment (manovotn) [reply docs/reference/src/main/asciidoc/environments.asciidoc]: Sure, that's a typo, meant to be `this way`, thanks.

**Pull request merged**